Avoid deprecation warning on explicit eta-expansion of nullary method

som-snytt opened the pull request:
Scala 2.12.4 produces a false deprecation warning for explicit eta-expansion of nullary methods:
```
scala> def f() = 1
scala> f _
<console>:13: warning: Eta-expansion of zero-argument method values is deprecated. Did you intend to write f()?
       f _
```

This PR fixes that.

Nuances https://github.com/scala/scala/commit/c15b9b730e76f2c33399ce4ffc2ffb3c71025531 until method value syntax is deprecated entirely.

Fixes scala/bug#10612

som-snytt committed:
Adapt method values more cannily, part 1

Introduce MethodValue to extract and build trees like `f _`.

retronym commented on src/reflect/scala/reflect/internal/Trees.scala:
```diff
@@ -473,6 +473,13 @@ trait Trees extends api.Trees {
+      case _ => None
+    }
+  }
+
```
Let's move `def makeMethodValue` into `MethodValue.apply`.

som-snytt replied:
OK.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5291,12 +5295,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           //
-          // Note: apparently `Function(Nil, EmptyTree)` is the secret parser marker
+          // Note: `Function(Nil, EmptyTree)` is the secret parser marker
           // which means trailing underscore -- denoting a method value. See makeMethodValue in TreeBuilder.
```
This comment could be removed now IMO, perhaps replaced with a one line doc comment on `MethodValue` to show the syntactic form it represents.

som-snytt replied:
Yes.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -909,8 +913,8 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+          def isExplicitEtaExpansion = original0 match {
+            case MethodValue(_) => true // tree shape for `f _`
             case _ => false
           }
```
We could move this right up to the first use of `MethodValue.unapply`

```scala
val isExplicitEtaExpansion = original0 != original
```

som-snytt replied:
Yes.

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5291,12 +5295,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
             typed1(suppressMacroExpansion(expr), mode, pt) match {
               case macroDef if treeInfo.isMacroApplication(macroDef) => MacroEtaError(macroDef)
-              case methodValue                                       => typedEta(checkDead(methodValue), expr)
+              case methodValue                                       => typedEta(checkDead(methodValue), tree)
```
So this is crux of the change. Can you describe how the surrounding changes allowed you go back to passing the entire `Typed` tree as `original` here, without incurring the infinite loop I was firefighting in c15b9b730e76f2c33399ce4ffc2ffb3c71025531?

som-snytt replied:
I probably can after I look at the code again.

lrytz commented:
@retronym this seems to be ready

SethTisue commented:
Cannily, isn't that where Puff the Magic Dragon lived?

som-snytt commented:
"I'll have what he's having."

SethTisue commented:
marking as "prio:blocker" because a lot of people have been asking about it

som-snytt commented:
I just broke out in a cold sweat.

adriaanm committed:
Method values are always eta-expanded

`m _` is shorthand for eta-expansion. Add an attachment to `m` in order
to know, in adapt, that it is used as an explicit method value.

adriaanm commented:
I reworked part 2 a bit -- no 💦

adriaanm commented:
Problem in neg/t10279.scala

adriaanm commented:
Ok, well, I was a little too hasty here -- sorry! Feel free to push -f your branch to its original state (https://github.com/som-snytt/scala/commit/933de7db5d5339ffc4cd085aab8022496e26e3f8). I was trying to move us away from special casing the treatment of `original` in `adapt`, because it should be a rare case and `adapt` is a core type checker method. I'll take another look on Monday

lrytz committed:
Some comments about re-typing implicit applications

lrytz commented:
@som-snytt yet another guy pushing to your branch, sorry... I pair-programmed with @adriaanm on this today, here's what we've come up with. I hope this looks good for 2.12.5.

@SethTisue since you were interested in this issue recently, maybe you'd like to check the test cases to see if everything looks right to you.

adriaanm commented on test/files/neg/t10279.scala:
```diff
@@ -1,13 +1,18 @@
+  val t5a: collection.immutable.WrappedString = t5 // don't ask me why
 
-  val bar = foo(1) _
+  def bar(i: Int)(implicit i: Int): Int = 0
```
Typo: rename second `i` argument

lrytz replied:
fixed...

adriaanm reviewed (changes requested):
Very nice! One typo in a test that would be good to fix.

adriaanm reviewed (approved):
💯 great cleanup, exhausti[ve/ng] tests ;-)

som-snytt commented:
I always wondered what @lrytz means when says "this seems to be ready." I kind of miss `LGTM`.

Also becoming very attached to attachments. They're like post-it notes.

SethTisue commented:
I'll merge this once Jenkins passes, and then that merge commit will be our RC that I'll send through the gauntlet of the community build

SethTisue commented:
/synch

hrhino commented:
🎉!

som-snytt commented:
@adriaanm missed opportunity to rename both attachments for tagging `case x @ _` and `f _` to `UnderscoreAttachment`.
